src/ymfm_adpcm.cpp

```diff
@@ -613,67 +609,78 @@ void adpcm_b_channel::write(uint32_t regnum, uint8_t value)
 	// dummy read counter
 	if (regnum == 0x00)
 	{
-		if (m_regs.execute())
+		// reset flag stops playback and holds output, but does not clear the
+		// externally-visible playing flag
+		if (m_regs.resetflag())
+			set_reset_status(STATUS_BRDY | (((m_status & STATUS_INTERNAL_PLAYING) != 0) ? STATUS_EOS : 0), STATUS_INTERNAL_PLAYING);
+
+		// all other modes set up for an operation
+		else
 		{
-			load_start();
-
-			// don't log masked channels
-			if ((debug::GLOBAL_ADPCM_B_CHANNEL_MASK & 1) != 0)
-				debug::log_keyon("KeyOn ADPCM-B: rep=%d spk=%d pan=%d%d dac=%d 8b=%d rom=%d ext=%d rec=%d start=%04X end=%04X pre=%04X dn=%04X lvl=%02X lim=%04X\n",
-					m_regs.repeat(),
-					m_regs.speaker(),
-					m_regs.pan_left(),
-					m_regs.pan_right(),
-					m_regs.dac_enable(),
-					m_regs.dram_8bit(),
-					m_regs.rom_ram(),
-					m_regs.external(),
-					m_regs.record(),
-					m_regs.start(),
-					m_regs.end(),
-					m_regs.prescale(),
-					m_regs.delta_n(),
-					m_regs.level(),
-					m_regs.limit());
+			// initialize the core state; appears to leave EOS flag alone until next execute
+			set_reset_status(STATUS_BRDY, STATUS_PLAYING | STATUS_INTERNAL_DRAIN | STATUS_INTERNAL_PLAYING);
+			m_curaddress = m_regs.external() ? (m_regs.start() << address_shift()) : 0;
+			m_buffer = (m_regs.cpudata() << 24) | (m_regs.cpudata() << 16);
+			m_nibbles = 4;
+			m_position = 0;
+			m_accumulator = 0;
+			m_adpcm_step = STEP_MIN;
+			m_output = 0;
+			m_prev_output = 0;
+
+			// if playing, set the playing status
+			if (m_regs.execute())
+			{
+				set_reset_status(STATUS_PLAYING | STATUS_INTERNAL_PLAYING, STATUS_EOS);
+				m_nibbles = 0;
+
+				// don't log masked channels
+				if ((debug::GLOBAL_ADPCM_B_CHANNEL_MASK & 1) != 0)
+					debug::log_keyon("KeyOn ADPCM-B: rep=%d spk=%d pan=%d%d dac=%d 8b=%d rom=%d ext=%d rec=%d start=%04X end=%04X pre=%04X dn=%04X lvl=%02X lim=%04X\n",
+						m_regs.repeat(),
+						m_regs.speaker(),
+						m_regs.pan_left(),
+						m_regs.pan_right(),
+						m_regs.dac_enable(),
+						m_regs.dram_8bit(),
+						m_regs.rom_ram(),
+						m_regs.external(),
+						m_regs.record(),
+						m_regs.start(),
+						m_regs.end(),
+						m_regs.prescale(),
+						m_regs.delta_n(),
+						m_regs.level(),
+						m_regs.limit());
+			}
 		}
-		else
-			m_status &= ~STATUS_EOS;
-		if (m_regs.resetflag())
-			reset();
-		if (m_regs.external())
-			m_dummy_read = 2;
+	}
+
+	// register 2-3 writes update the current address if configured for external
+	else if ((regnum & ~1) == 2 && m_regs.external())
+	{
+		m_curaddress = m_regs.start() << address_shift();
+		m_buffer = (m_regs.cpudata() << 24) | (m_regs.cpudata() << 16);
+		m_nibbles = 4;
 	}
 
 	// register 8 writes over the bus under some conditions
 	else if (regnum == 0x08)
 	{
 		// if writing from the CPU during execute, clear the ready flag
 		if (m_regs.execute() && !m_regs.record() && !m_regs.external())
-			m_status &= ~STATUS_BRDY;
+			set_reset_status(0, STATUS_BRDY);
```

Comment:
I have not checked this behavior of BRDY.

What I confirmed was that 
writing 0x80 into REG10 during playback clears BRDY status and it keeps "0" .
Maybe nobody cares...

* BRDY is "1"
* write 0xa0 to REG00 -- start playback
* write 0x80 to REG10 -- BRDY becomes "0"
* write 0xa1 to REG00 -- stop playback, BRDY keeps "0"
* write 0x80 to REG10 -- BRDY keeps "0"
* write 0x00 to REG00 -- BRDY becomes "1"